### PR TITLE
[REDO] Add torch.sparse_coo_tensor factory.

### PR DESCRIPTION
**Pull request by gchanan**
Notes:
1) I didn't add Tensor.new_sparse_coo_tensor; it didn't seem particularly useful, but it's easy to add
2) This doesn't do the type inference, i.e. torch.sparse_coo_tensor(indices=LongTensor, values=IntTensor)
will return a sparse tensor corresponding to the default type rather than a sparse IntTensor.  We can add
type inference later when we add it to other factories.

### CODE REVIEW

**Comment by gchanan** [comment]
@colesbury can you take a look at the latest commit; it moves the dispatch_type from python_variable_methods and uses it more generally.

File: tools/autograd/templates/python_torch_functions.cpp [hunk 1]
```diff
@@ -120,9 +127,10 @@ static PyMethodDef torch_functions[] = {
   {"dsmm", (PyCFunction)THPVariable_mm, METH_VARARGS | METH_KEYWORDS | METH_STATIC, NULL},
   {"from_numpy", (PyCFunction)THPVariable_from_numpy, METH_STATIC | METH_O, NULL},
   {"hsmm", (PyCFunction)THPVariable_hspmm, METH_VARARGS | METH_KEYWORDS | METH_STATIC, NULL},
+  {"_promote_types", (PyCFunction)THPVariable__promote_types, METH_VARARGS | METH_KEYWORDS | METH_STATIC, NULL},
   {"saddmm", (PyCFunction)THPVariable_sspaddmm, METH_VARARGS | METH_KEYWORDS | METH_STATIC, NULL},
+  {"sparse_coo_tensor", (PyCFunction)THPVariable_sparse_coo_tensor, METH_VARARGS | METH_KEYWORDS | METH_STATIC, NULL},
```
Comment (colesbury) [inline tools/autograd/templates/python_torch_functions.cpp]: Should probably document this sooner rather than later

**Review by colesbury — APPROVED** [review]

**Comment by gchanan** [comment]
onnx-fb-universe seems borked on this PR, but this shouldn't tourch onnx so I'll merge.

File: torch/csrc/utils/tensor_conversion_dispatch.cpp [hunk 2]
```diff
@@ -0,0 +1,50 @@
+#pragma once
```
Comment (jamesr66a) [inline torch/csrc/utils/tensor_conversion_dispatch.cpp]: I don't think this should be here

**Pull request merged**